crates/node/builder/src/launch/common.rs

```diff
@@ -501,20 +502,50 @@ where
         )?
         .with_prune_modes(self.prune_modes());
 
-        // Check for consistency between database and static files. If it fails, it unwinds to
-        // the first block that's consistent between database and static files.
-        if let Some(unwind_target) =
-            factory.static_file_provider().check_consistency(&factory.provider()?)?
-        {
+        // Check for consistency between database, static files, and RocksDB. If any
+        // inconsistencies are found, unwind to the first block that's consistent across all
+        // storage layers.
+        //
+        // Static file check runs first since RocksDB pruning logic may depend on static file data.
+        // We compute a combined unwind target from both checks and run a single unwind pass.
+        let static_file_unwind = factory
+            .static_file_provider()
+            .check_consistency(&factory.provider()?)?
+            .map(|target| match target {
+                PipelineTarget::Unwind(block) => block,
+                PipelineTarget::Sync(_) => unreachable!("check_consistency returns Unwind"),
+            });
+
+        // RocksDB consistency check - runs after static files since it may use static file data
+        // for pruning transaction hashes.
+        let rocksdb_unwind =
+            factory.rocksdb_provider().check_consistency(&factory.database_provider_ro()?)?;
+
+        // Combine unwind targets - take the minimum (most conservative) if both exist
+        let unwind_target = match (static_file_unwind, rocksdb_unwind) {
+            (None, None) => None,
+            (Some(a), None) | (None, Some(a)) => Some(a),
+            (Some(a), Some(b)) => Some(a.min(b)),
+        };
+
+        if let Some(unwind_block) = unwind_target {
             // Highly unlikely to happen, and given its destructive nature, it's better to panic
             // instead.
+            let inconsistency_source = match (static_file_unwind, rocksdb_unwind) {
+                (Some(_), Some(_)) => "static file <> database and RocksDB <> database",
+                (Some(_), None) => "static file <> database",
+                (None, Some(_)) => "RocksDB <> database",
+                (None, None) => unreachable!(),
+            };
             assert_ne!(
```

Comment:
this will panic when rocksdb returns `Some(0)` as unwind target, which can happen for instance when an existing node enables rocksdb flags. is this what we want? this will be a common case for nodes migrating to using rocksdb, maybe the message in the panic is not clear enough, or we want to let them unwind to 0 given that they will need to resync in any case, correct?

also, i see the comment "Highly unlikely to happen, and given its destructive nature, it's better to panic" given that it is more likely to happen now, if we want this behavior we should update the comment, wdyt?